[ng] router chapter

chalin opened the pull request:
- Ignore the `angular/guide/router.html` as that is obsolete (it will be removed shortly, after I've finished extracting what is still useful from the file).

Contributes to #198

kwalrath commented on src/_data/side-nav.yml:
```diff
@@ -51,6 +51,19 @@
         - title: 6. HTTP
           permalink: /angular/tutorial/toh-pt6
 
+    - title: Router
```
It seems weird to have Router at the same level as Guide, Tutorial, and Advanced. Although... it might be defensible if we put Angular-related packages (such as angular2_components) in the same place, since the router will also be separate (right?). But maybe even then, they should go under WEB LIBRARIES?

For now, I'm inclined to have the first page (Routing & Navigation) either in the guide, after Template Syntax, or in Advanced. The remaining pages would be in Advanced, grouped together somehow so that you don't see all of them unless you're interested.

chalin replied:
The side nav depth is limited to 2 levels (+ a possible section heading) for good reason, but it does force us to give some extra though to the docs organization.

Yes, we will inevitably want a way to be able to document angular relevant libraries and packages separately, if only because we cannot fit pack it all into two levels of the side nav.

I think that a Router/Routing section fits well where it is place, though maybe I can move it to after "Advanced".

kwalrath commented on src/angular/router/index.jade:
```diff
@@ -0,0 +1,312 @@
+---
+layout: angular
+title: Angular Router
```
Why call this "Angular Router" instead of "Routing & Navigation" or just "Routing"?

chalin replied:
I was looking for something short. "Routing" is a good candidate, though we cover more than routing, this new guide describes the router features.

chalin replied:
Changed to "Routing".

kwalrath commented on src/angular/router/index.jade:
```diff
@@ -0,0 +1,312 @@
+---
+layout: angular
+title: Angular Router
+description: Discover the basics of screen navigation with the Angular Router
```
Why "Angular Router" instead of "Angular router"?

chalin replied:
"Router" is taken as a proper name (as in "the Router library"). This is who it was written in the TS docs and, as I mentioned elsewhere, I chose to focus on getting core content done, thinking that we can come back to tweak capitalization as needed.

chalin replied:
fixed

kwalrath commented on src/angular/router/index.jade:
```diff
@@ -0,0 +1,312 @@
+title: Angular Router
+description: Discover the basics of screen navigation with the Angular Router
+nextpage:
+  title: 1. Crisis Center App Starter
```
I know there's precedent, but I dislike having numbers in titles unless the number really matters and is labeled (e.g. "Step 1").

chalin replied:
I agree. The titles used to be "Milestone 1. ..." but at the same time we want to keep titles short. I.e., there is a balance to be had; cf. #175.

kwalrath commented on src/_data/side-nav.yml:
```diff
@@ -51,6 +51,19 @@
+        - title: Introduction
+          permalink: /angular/router
+          urlExactMatch: true
+        - title: 1. App Starter
```
We should name pages after user tasks, rather than the examples they feature.

chalin replied:
Done

kwalrath commented on src/angular/router/crisis-center-1.jade:
```diff
@@ -0,0 +1,294 @@
+description: Getting started with the router
+prevpage:
+  title: Angular Router
+  url: /angular/router/index
```
Should `index` really be part of the URL?

kwalrath commented on src/angular/router/index.jade:
```diff
@@ -0,0 +1,312 @@
+//- TODO: sweep this chapter and replace "[an] array" by "[a] list".
+
+:marked
+  The Angular **Router** enables navigation from one [view](../glossary.html#view) to the next
```
Why the "Angular..." prefix, and why capital R Router?

kwalrath replied:
(This is a global comment.)

chalin commented on src/angular/router/index.jade:
```diff
@@ -0,0 +1,312 @@
+:marked
+  ### Configuration
+
+  When the browser's URL changes, the router looks for a corresponding **`RouteDefinition`**
```
I believe that most of the time bold is used to mark the point of first use of the named class. Do you still want me to remove the bold (this convention is also used in other chapters).

chalin commented on src/angular/router/index.jade:
```diff
@@ -0,0 +1,312 @@
+:marked
+  ### Router outlet
+
+  Given this configuration, when the browser URL for this application becomes `/heroes`,
```
Fixed

chalin commented on src/angular/router/index.jade:
```diff
@@ -0,0 +1,312 @@
+      You can bind that list to a <code>RouterLink</code> or pass the list as an argument to
+      the <code>Router.navigate</code> method.
+  tr
+    td <b><i>Routing Component</i></b>
```
Agreed

chalin replied:
I've switched to sentence case.

chalin commented on src/angular/router/index.jade:
```diff
@@ -0,0 +1,312 @@
+
+.l-main-section
+:marked
+  ## The Sample Application
```
fixed

chalin commented on src/angular/router/index.jade:
```diff
@@ -0,0 +1,312 @@
+  1. A *Heroes* area for maintaining the list of heroes employed by the agency.
+  1. An *Admin* area to manage the list of crises and heroes.
+
+  Try it by clicking on this <live-example title="Hero Employment Agency Live Example">live example link</live-example>.
```
Ok

chalin commented on src/angular/router/index.jade:
```diff
@@ -0,0 +1,312 @@
+      a route. Clicking an element with a <code>routerLink</code> directive
+      that is bound to a <i>link parameters list</i> triggers a navigation.
+  tr
+    td <b><i>Link Parameters List</i></b>
```
I've switched to sentence case.

chalin commented on src/angular/router/index.jade:
```diff
@@ -0,0 +1,312 @@
+  The shell component has a `RouterOutlet` where it can display views produced by the router.
+  It has `RouterLink`s that users can click to navigate via the router.
+
+  Here are the key `Router` terms and their meanings:
```
Changed to
> router

chalin commented on src/angular/router/index.jade:
```diff
@@ -0,0 +1,312 @@
+  ### Router outlet
+
+  Given this configuration, when the browser URL for this application becomes `/heroes`,
+  the router matches that URL to the `RouteDefinition` named `Heroes` and displays the `HeroesComponent`
```
Opened https://github.com/dart-lang/site-webdev/issues/284

chalin commented:
@matanlurey @kwalrath - this is ready for review. This first version of the page is mainly based on the TS router page _before_ they migrated to the new (which is current) router API. The material for Milestone 6 is still TBC, but most of the previous material is complete. If you are agreeable to the current version of the page, we can make iterative improvements (in particular for Milestone 6).

kwalrath commented:
Taking off Matan, since we've requested reviews from @henrysenger and @potato620.

henrysenger commented on src/angular/guide/router/crisis-center-1.jade:
```diff
@@ -0,0 +1,225 @@
+  You must add a [&lt;base href> element](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/base)
+  to the app's `index.html` for `pushState` routing to work.
+  The browser uses the base `href` value to prefix *relative* URLs when referencing
+  CSS files, scripts, and images.
```
Consider giving some examples here of how the router uses the base href. In the current documentation for the existing routers it isn't clear here what specifically changing the base href will do.

henrysenger commented on src/angular/guide/router/crisis-center-1.jade:
```diff
@@ -0,0 +1,225 @@
+  The `@RouteConfig` annotation combines the creation of a new router, its configuration, and
+  its assignment to a host component in a single step.
+
++makeExcerpt('app/app.component.1.ts', 'routes', '')
```
Probably a stupid question but why are we referencing Typescript files here? I thought this was the Dart documentation?

chalin replied:
Most doc pages use a form of "subclassing" to create the Dart docs. That is, the base class is the TS doc and we just write a minimal Dart doc subclass. Under this scheme, the Jade mixin (`+makeExcerpt`) handles the path translation from TS to Dart. This is part of what makes the subclassing scheme work. (The mixin does other work like deriving the title from the path and dropping file variant numeric suffixes in the process.)

Just like in programming, subclassing reduces the effort of creating the Dart-side docs. But since the split we've been breaking away from this scheme. So eventually, TS example paths and other such artifacts from times past, will be cleaned out/replaced with Dart-only prose elements.

henrysenger commented on src/angular/guide/router/crisis-center-1.jade:
```diff
@@ -0,0 +1,225 @@
+  the router activates an instance of the `CrisisCenterComponent` and displays its view.
+
+  - When the application requests navigation to the route named `CrisisCenter`, then the router
+  activates an instance of `CrisisCenterComponent`, displays its view, and updates the
```
What does "activates" mean exactly here? Will a new CrisisCenterComponent be instantiated every time I navigate to that URL? If I leave and come back, will the same component be reused?

henrysenger commented on src/angular/guide/router/crisis-center-1.jade:
```diff
@@ -0,0 +1,225 @@
++makeExcerpt('app/app.component.1.ts', 'routes', '')
+
+:marked
+  The `@RouteConfig()` is applied to `AppComponent` which makes it the router's host component.
```
Consider stating that the host component must contain a router outlet.

henrysenger commented on src/angular/guide/router/crisis-center-1.jade:
```diff
@@ -0,0 +1,225 @@
+//- TODO: mention routerLink: class.router-link-active here?
+
+:marked
+  ## *Declaring router providers and directives*
```
Consider moving this section up.

henrysenger commented on src/angular/guide/router/crisis-center-2.jade:
```diff
@@ -0,0 +1,121 @@
+:marked
+  ## Redirect route
+
+  As an alterivate solution, remove the `useAsDefault` argument that you just added,
```
s/alterivate/alternative/

chalin replied:
lol, wow, thanks for spotting that.
Fixed

henrysenger commented on src/angular/guide/router/crisis-center-2.jade:
```diff
@@ -0,0 +1,121 @@
+
+  As an alterivate solution, remove the `useAsDefault` argument that you just added,
+  and instead add a `Redirect` route that will translate an initial relative URL (`''`)
+  to the desired default path (`/heroes`).
```
Do redirects happen with a pushState or replaceState?

henrysenger commented on src/angular/guide/router/crisis-center-2.jade:
```diff
@@ -0,0 +1,121 @@
+    By convention, the path `/**` is used, resulting in a path segment name of just `*`.
+
+:marked
+  The router will select a wildcard route if it can't match a route earlier in the configuration.
```
Maybe too in-depth for this tutorial, but consider explaining how the router must match the *entire* path (and not a prefix of it) in order for it to count as a match.

kwalrath commented:
Thanks for the review, @henrysenger. We expect @chalin to be able to work on this doc next week.

potato620 commented on src/angular/guide/router/crisis-center-3.jade:
```diff
@@ -0,0 +1,368 @@
+  img(src='/resources/images/devguide/router/selected-hero.png' alt="Selected hero")
+
+:marked
+  This becomes possible if you can include the viewed hero's `id` in the URL when
```
mention somewhere that an optional parameter is a query param and link to an online resource

potato620 commented on src/angular/guide/router/crisis-center-3.jade:
```diff
@@ -0,0 +1,368 @@
+
+.l-sub-section
+  :marked
+    The router can also encode parameters using the 
```
IMO seems unnecessary to mention this here

potato620 commented on src/angular/guide/router/crisis-center-4.jade:
```diff
@@ -0,0 +1,449 @@
+  it has no business logic, and it's template has no links, just a title and 
+  `<router-outlet>` for the crisis center child views.
+
+  Unlike `AppComponent`, and most other components, it _lacks a selector_.
```
There is still a selector ATM

chalin replied:
Yes, there is still a selector, but selector's should not be mandatory, so I'd rather keep the prose as is for now and leave the warning in the code excerpt, directing the reader to  https://github.com/dart-lang/angular2/issues/60.

potato620 commented on src/angular/guide/router/crisis-center-4.jade:
```diff
@@ -0,0 +1,449 @@
+  These two routes navigate to the crisis center child components,
+  `CrisisCenterHomeComponent` and `CrisisDetailComponent`, respectively.
+
+  Being _child routes_, there are some *important differences* in the way the router treats them:
```
"important differences", but only one is listed. Maybe also mention that the paths are relative to the CrisisCenterComponent

ie: the /:id is actually /crisis-center/:id

chalin replied:
There are actually two bullets (though that is a bit harder to see from the .jade source), and the second bullet kind of makes the point you raised:
> ![screen shot 2017-02-07 at 2 29 43 pm](https://cloud.githubusercontent.com/assets/4140793/22714577/f96dd2c2-ed41-11e6-80b1-018633c993db.png)

kwalrath commented:
Thanks for the review, @potato620. I'll leave this to @chalin to answer when he gets back.

chalin commented on src/angular/guide/router/index.jade:
```diff
@@ -0,0 +1,315 @@
+  It is an optional service packaged within its own library, `angular2.router`.
+  Import it like any other package.
+
++makeExcerpt('lib/app_component_1.dart', 'import', '')
```
By default `+makeExcerpt` creates a code excerpt with a title (which matches the first argument), a blank 3rd argument instructs the mixin to drop the title, and therefore only show the code.

chalin commented on src/_data/side-nav.yml:
```diff
@@ -67,8 +67,18 @@
+            - title: "Routing 2: Hero Feature"
+              permalink: /angular/guide/router/crisis-center-2
+            - title: "Routing 3: Crisis Center Feature"
+              permalink: /angular/guide/router/crisis-center-3
```
Actually, none of those sub-sub-menu items should be there because the side nav doesn't show entries to that depth.
Fixed

chalin commented on src/angular/guide/router/index.jade:
```diff
@@ -0,0 +1,315 @@
+  adding its routes using a **`@RouteConfig`**
+  applied to the router's host component:
+
++makeExcerpt('app/app.component.1.ts (routes)', 'app-component-routes')
```
I'm not sure I understand this request. Are you asking for a hyperlink to the file? We don't actually publish such transitory files -- i.e., files with numbers in their names.  Such files _are_ accessible via the docs repo, but we don't link to them, instead we direct readers to the official samples collected under https://github.com/angular-examples.

kwalrath replied:
If you want to see the source code, the best way is probably to go to the rendered page and search for "(routes)".

@chalin, where is the rendered page?

chalin replied:
Sorry, were you asking that a link be added in the prose, or that you simply wanted to see this transient file's content? If the latter, then see
- https://github.com/dart-lang/site-webdev/pull/283/files#diff-d8a501436c8079db9f220c459130a937 or
- https://github.com/chalin/site-webdev/blob/e1be6e9472bb346464e91b409fa10d69ba868594/public/docs/_examples/router/dart/lib/app_component_1.dart

chalin commented on src/angular/guide/router/crisis-center-3.jade:
```diff
@@ -0,0 +1,368 @@
+  - Add the `HeroService` to the `providers` list of `AppComponent`
+    so that it is available everywhere in the app. This should be the result:
+
++makeExcerpt('lib/app_component_3.dart (v1)')
```
The `v1` names the region being shown here. This could also have been written as:
`+makeExcerpt('lib/app_component_3.dart', 'v1')`.

chalin commented on src/angular/guide/router/crisis-center-3.jade:
```diff
@@ -0,0 +1,368 @@
+  that creates a slot in the path for a **Route Parameter**.
+  In this case, the router will insert the `id` of a hero into that slot.
+
+  If you tell the router to navigate to the detail component and display "Magneta", 
```
done

kwalrath commented:
@chalin I've staged this locally and am looking at the sample, and I noticed a few issues:

1. The text and image show Heroes as being the first screen, but it displays Crisis Center.

![image](https://cloud.githubusercontent.com/assets/2164483/22720780/c2779986-ed61-11e6-979d-e352fd6c39e1.png)

2. As the screenshot shows, I see docregion comments in the UI (in both the crisis and hero screens).

3. The text says that selecting a hero takes you to an editing screen, but it doesn't.

4. There's no indication, from the buttons, of which screen you're looking at. The screenshot, on the other hand, implies that the current screen's button is highlighted.

5. The UI of the crisis editor has an extra button not shown in the screenshot. Also, the order of the buttons is different.

6. The text implies that you can see the crisis list at the same time as the crisis editor, but you can't.

7. http://localhost:4001/angular/guide/router/crisis-center-6 has **[TBC]** at the end. What should it really have?

I'll take a more careful look at the words tomorrow.

chalin commented:
> 1. The text and image show Heroes as being the first screen, but it displays Crisis Center.

Fixed prose and code (that behavior was from the old version of the AngularDart router draft).

> 2. As the screenshot shows, I see docregion comments in the UI (in both the crisis and hero screens).

Those don't appear in the [published example](https://angular-examples.github.io/router), only when building and running from the "raw" sources contained under webdev. But when a reader is walking through the material he/she won't be typing docregion tags and so it won't appear for them either.
> 3. The text says that selecting a hero takes you to an editing screen, but it doesn't.

Fixed (this was due to test code I had added for the next parts).

> 4. There's no indication, from the buttons, of which screen you're looking at. The screenshot, on the other hand, implies that the current screen's button is highlighted.

Correct, that is not implemented yet. It is part of my to do list to add in the next round of changes.

Your other points (5, 6, & 7) are, as indicated TBC.  As I pointed out before, this is a _very_ long guide and I'd much prefer adopting an iterative approach to its publication. Part of the challenge is that features are handled differently in NgDart as compared to NgTS.

kwalrath commented:
> Your other points (5, 6, & 7) are, as indicated TBC. As I pointed out before, this is a very long guide and I'd much prefer adopting an iterative approach to its publication. Part of the challenge is that features are handled differently in NgDart as compared to NgTS.

I'm fine with an iterative approach, but we need to be clear to readers about what they can expect. If they find errors and inconsistencies, they won't know what parts of the doc, if any, they can trust. The clearer we can be, the better.

chalin commented:
For we can mark the entire router guide as having a "DRAFT" status. On my next pass, I can take note of which parts are complete and which parts have known issues (and then identifying the specific issues in the text via, say, callouts). WDYT?

kwalrath commented:
After a few small changes, I'm fine with releasing this as a draft.

1. Add a prominent note to /angular/guide/router that this is a draft, perhaps soliciting feedback. Maybe, in a green(?) box:
   
   **This is a draft.** The router pages are still being updated. We've published them so you can get the information you need and [give us feedback](https://github.com/dart-lang/site-webdev/issues/new).
 
1. /angular/guide/router should point to /angular/guide/pipes as the previous page. (Pipes correctly has router as the next page.)

1. Change the "[TBC]" in crisis-center-6 to something a little clearer. Perhaps just **Under construction...**

chalin commented:
@kwalrath - I've addressed the items you mentioned in the previous comment.

kwalrath commented:
LGTM. We'll iterate!

chalin commented:
@henrysenger @potato620 - just to let you know that I addressed your most pressing concerns in this PR. The other points that you raised will be addressed in followup PR.

potato620 commented:
sgtm 👍🏻
